src/comps/qb.cc

```diff
@@ -0,0 +1,194 @@
+/*
+TODO #1: Update implementation so that no copy of the original data is needed.
+
+TODO #3: Need a test case with switching between different orthogonalization types
+
+On early termination, data in B is moved, but not sized down
+
+TODO: If user gives tol 0, change it to machine epsilon * 100
+*/
+
+#include <RandBLAS.hh>
+#include <lapack.hh>
+#include <RandLAPACK.hh>
+
+#include <math.h>
+
+#define ORTHONORM_CHECKS
+#define COND_CHECK
+#define VERBOSE
+#define USE_QR
+
+namespace RandLAPACK::comps::qb {
+
+template <typename T>
+void QB<T>::QB2(
+    int64_t m,
+    int64_t n,
+    std::vector<T>& A,
+    int64_t& k,
+    int64_t block_sz,
+    T tol,
+    std::vector<T>& Q,
+    std::vector<T>& B
+){
+    using namespace blas;
+    using namespace lapack;
+
+    int64_t curr_sz = 0;
+    int64_t next_sz = 0;
+
+    T* A_dat = A.data();
+    // pre-compute nrom
+    T norm_A = lange(Norm::Fro, m, n, A_dat, m);
+    // Immediate termination criteria
+    if(norm_A == 0.0)
+    {
+        // Zero matrix termination
+        k = curr_sz;
+        this->termination = 1;
+        return;
+    }
+    
+    // If the space allocated for col in Q and row in B is insufficient for any iterations ...
+    if(std::max( Q.size() / m, B.size() / n) < k)
+    {
+        // ... allocate more!
+        this->curr_lim = std::min(4 * block_sz, k);
+        // No need for data movement in this case
+        RandLAPACK::comps::util::resize<T>(m * this->curr_lim, Q);
+        RandLAPACK::comps::util::resize<T>(this->curr_lim * n, B);
+    }
+    else
+    {
+        this->curr_lim = k;
+    }
+    
+    // Copy the initial data to avoid unwanted modification TODO #1
+    std::vector<T> A_cpy (m * n, 0.0);
+    T* A_cpy_dat = A_cpy.data();
+    lacpy(MatrixType::General, m, n, A_dat, m, A_cpy_dat, m);
+
+    T norm_B = 0.0;
+    T prev_err = 0.0;
+    T approx_err = 0.0;
+
+    T* Q_gram_dat;
+    T* Q_i_gram_dat;
+
+    if(this->orth_check)
+    {
+        Q_gram_dat = RandLAPACK::comps::util::resize(this->curr_lim * this->curr_lim, this->Q_gram);
+        Q_i_gram_dat = RandLAPACK::comps::util::resize(block_sz * block_sz, this->Q_i_gram);
+    }
+
+    T* QtQi_dat = RandLAPACK::comps::util::resize(this->curr_lim * block_sz, this->QtQi);
+    T* Q_i_dat = RandLAPACK::comps::util::resize(m * block_sz, this->Q_i);
+    T* B_i_dat = RandLAPACK::comps::util::resize(block_sz * n, this->B_i);
```

Comment:
These calls to resize are really verbose. Is there a C++ equivalent of a "Pythonic" ``import randlapack.comps.util as rlautil``?